### ISSUE TITLE
Wrong dependencies for python-pip

### ISSUE THREAD

**Issue by Paebbels** [msys2/MINGW-packages#7580]
Since when does python-pip depend on so many Python packages?

```python
depends=("${MINGW_PACKAGE_PREFIX}-python-appdirs
         "${MINGW_PACKAGE_PREFIX}-python-cachecontrol"
         "${MINGW_PACKAGE_PREFIX}-python-colorama"
         "${MINGW_PACKAGE_PREFIX}-python-contextlib2"
         "${MINGW_PACKAGE_PREFIX}-python-distlib"
         "${MINGW_PACKAGE_PREFIX}-python-html5lib"
         "${MINGW_PACKAGE_PREFIX}-python-lockfile"
         "${MINGW_PACKAGE_PREFIX}-python-msgpack"
         "${MINGW_PACKAGE_PREFIX}-python-packaging"
         "${MINGW_PACKAGE_PREFIX}-python-pep517"
         "${MINGW_PACKAGE_PREFIX}-python-progress"
         "${MINGW_PACKAGE_PREFIX}-python-pyparsing"
         "${MINGW_PACKAGE_PREFIX}-python-pytoml"
         "${MINGW_PACKAGE_PREFIX}-python-requests"
         "${MINGW_PACKAGE_PREFIX}-python-retrying"
         "${MINGW_PACKAGE_PREFIX}-python-six"
         "${MINGW_PACKAGE_PREFIX}-python-webencodings")
```
Source: https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-python-pip/PKGBUILD#L15-L31

When I install pip in e.g. a Debian or a Windows, only `colorama` is installed. Not even `wheel` is installed, even when nowadays most packages are only delivered in wheel format.

Can you please explain, why a pip in MinGW needs so many dependencies it's not using at all?

-----------------------------
When checking the sources of [pip](https://github.com/pypa/pip), there is no requirement specified, not even `colorama`.

**Comment by Biswa96** [comment]
How many of them do you find here https://github.com/archlinux/svntogit-packages/blob/packages/python-pip/trunk/PKGBUILD ?

**Comment by Paebbels** [comment]
@Biswa96 Arch Linux is not a reference (maybe an indication). Only the source code of a package define what is needed and was is unneeded. So if the source code of pip (see link above) doesn't define dependencies, why is MinGW defining so many dependencies?

If you're saying MinGW copied dependencies from Arch Linux, then the questions are:
* Why has Arch Linux created so many dependencies?
* Why does MinGW copy dependencies from Arch Linux?

**Comment by lazka** [comment]
Debian/Ubuntu includes all dependencies as wheels: https://packages.ubuntu.com/groovy/all/python-pip-whl/filelist

**Comment by Paebbels** [comment]
> Debian/Ubuntu includes all dependencies as wheels: https://packages.ubuntu.com/groovy/all/python-pip-whl/filelist

Ubuntu lists 28 packages, while Arch Linux and MinGW list 17 packages.  
Please note, your reference is only correct for Ubuntu, not for Debian.

See dependencies of Debian (Buster):  
![image](https://user-images.githubusercontent.com/956109/103531965-a70c8100-4e8a-11eb-8921-96587f458362.png)  
Source: https://packages.debian.org/buster/python3-pip

OK, let's take a docker container from DockerHub and start Python 3.8 (`docker run --rm -it python:3.8 /bin/bash`). Tag `python:3.8` is equal to `python:3.8-buster`, which is a Debian Buster based container extended by Python 3.8 (hash `c3a05daea4e7`). So let's run `pip list` inside and investigate installed packages...

![image](https://user-images.githubusercontent.com/956109/103532662-f1dac880-4e8b-11eb-947b-01d27aac8551.png)

As we can see:
* Less installed packages
* Packages installed that are not listed in MinGW

-----------
Again, dependencies are not defined by distributions, they are a requirements that the provided library or package can be executed without any loading or linking problems. Distributions can e.g. decide to add suggested packages and plugins, so a user has to reference less packets. This is usually done via virtual packages or bundling packages.

For now, it's hard to understand why MinGWs `python-pip` needs e.g. `toml` support, whereas e.g `wheel` is not included.

So that's why I'm asking: Why are there so many dependencies added and on the other hand really useful -- almost required -- dependencies like `wheel` are not included.

**Comment by naveen521kk** [comment]
Just saying if you look at the [tar file](https://files.pythonhosted.org/packages/ca/1e/d91d7aae44d00cd5001957a1473e4e4b7d1d0f072d1af7c34b5899c9ccdf/pip-20.3.3.tar.gz) of pip from PyPI, you could see a folder called `pip/_vendor` just looks what's over there. Similarly, go to `C:\msys64\mingw64\lib\python3.8\site-packages\pip` and check the `_vendor` directory. You would be seeing an empty one right? 

That's how it is in package managers, nothing should be vendored, everything should be from the repo. If you got the vendored dependency, you would be missing the patches from this repo, for those packages, and things would be really weird then.

**Comment by Paebbels** [comment]
@naveen521kk thanks for explain this and pointing to a list of embedded packages that get silently distributed and installed by pip and not registered in pip's package list.

So I understand now, distributions like MinGW are now overruling the intended packages by installing packages directly via their own package manager. Then I have these follow up questions:
* Does pip switch from local packages in `_vendor` to global package provided e.g. via pacman, when they get installed?
* pip is very strict about package compatibility. How do you ensure, packages installed via pacman don't break pip?  
  As additional packages are not known to pip, pip cannot check for version constrains, right?
* Do pacman and pip cooperate if I update a package either via pacman or via pip? Or do I break the other tools package databases with such operations?

```
appdirs==1.4.4
CacheControl==0.12.6
colorama==0.4.4
contextlib2==0.6.0.post1
distlib==0.3.1
distro==1.5.0
html5lib==1.1
ipaddress==1.0.23  # Only needed on 2.6 and 2.7
msgpack==1.0.0
packaging==20.8
pep517==0.9.1
progress==1.5
pyparsing==2.4.7
requests==2.25.0
    certifi==2020.11.08
    chardet==3.0.4
    idna==2.10
    urllib3==1.26.2
resolvelib==0.5.3
retrying==1.3.3
setuptools==44.0.0
six==1.15.0
toml==0.10.2
webencodings==0.5.1
```

**Comment by umarcor** [comment]
Coming from a recent conversation with @danyeaw in the gitter chat room, I find the last questions by @Paebbels interesting to be documented.

> [January 2, 2021 8:31 PM](https://gitter.im/msys2/msys2?at=5ff0ca2997312f4b6b052656)
> sometimes packages cannot be installed through pip because they don't work, and it's difficult to know that. Not all MSYS2 packaged Python modules need to be packaged for using them. Some of them are dependencies of others only.

**Comment by mingwandroid** [comment]
MSYS2/mingw-w64 python should never use PIP. Not for anything that pacman does at least. If our end-users want to play with PIP that is up to them.

The reason? There's no Python ABI string for MSYS2/mingw-w64 python and there needs to be before this can work. But that would involve upstreaming most if not all of our python patches which is, I fear, never going to happen. At least I will never have time to do it personally and I wrote half of them originally (if not more).

**Comment by mingwandroid** [comment]
In fact I'd rather remove PIP than give the impression that it's something that's expected to work well (or at all).

**Comment by Paebbels** [comment]
> In fact I'd rather remove PIP than give the impression that it's something that's expected to work well (or at all).

So we go back into stone age of computer science and distribute packages via floppy disks and printed manuals with installation instructions? You're the first person since 14 years of using Python who says pip is bad and should never be used at all. If this is true, why does the world work differently from what you propose?

For most packages it's completely irrelevant what underlying system is used as long as it is Python, because it's platform independent code !! Then there is a remaining tiny fraction of packages coming with platform dependent code. So here we can start to talk what's making problems.

The wheel format was invented to improve problems with platform specific code and binary extensions. I'm not involved in details here, but people created that format; it's now the defacto standard; and many packages with native extensions are distributed that way.

So if you have a problem and you claim Debian and others have this problem too, then -- damn -- stick your heads together with pip / wheel people and find a solution instead of building walls!

-------------
Anyhow, this issue started about dependencies of pip. As it was proven to me (thanks again for pointing to the right sources!), these dependencies are hidden dependencies and are not exposed to the user in a list of installed packages. Nonetheless, I still have some follow up questions posted in my last answer.

**Comment by mingwandroid** [comment]
> So we go back into stone age of computer science and distribute packages via floppy disks and printed manuals with installation instructions? You're the first person since 14 years of using Python who says pip is bad and should never be used at all. If this is true, why does the world work differently from what you propose?

How is removing PIP going back to the stone age? No, we use pacman, a *far* better more *general* language independent, well tested package manager instead, as @naveen521kk already said. I do not understand why pythonistas (apologies if you don't like being labelled as that) cling so dearly to PIP. A package manager that only got a real, actual solver in the last few months! Note I don't propose seriously removing it but it shouldnt be used on MSYS2 in general despite your liking for it,

> pip is very strict about package compatibility. How do you ensure, packages installed via pacman don't break pip? As additional packages are not known to pip, pip cannot check for version constrains, right?

Err, no, it is far from it. It is a train-wreck in this regard. See above where I mention solvers.

FWIW, pip requires none of these dependencies and should not list a single one IMHO. I have no idea why ArchLinux list 
them.

> The wheel format was invented to improve problems with platform specific code and binary extensions. I'm not involved in details here, but people created that format; it's now the defacto standard; and many packages with native extensions are distributed that way.

The wheel format specifies that all binaries are compiled against the MS UCRT. MinGW-w64 is compiled against MSVCRT. Please stop telling me things about how great things are (if I just ignore reality).

> So if you have a problem and you claim Debian and others have this problem too, then -- damn -- stick your heads together with pip / wheel people and find a solution instead of building walls!

I have 0 influence or interest in influence over Debian. I am simply stating a fact that anyone who works on python packaging in a real distro that I've spoken to (and that's a good few since I work in this field professionally) have a strong dislike for PIP and all the many missteps in its design (vendoring, static linking, no actual solver).

That you haven't heard anyone complain or that you like it is neither here nor there!

**Comment by danyeaw** [comment]
@mingwandroid We are not removing pip. I'm not sure why you are on here even suggesting this, that would break many many apps which depend on pip for packaging and releasing for Windows and would make python useless for most users.

If we don't require these many dependencies, someone please submit a PR to fix it.

**Comment by mingwandroid** [comment]
@danyeaw who is we in this case?

msys2 was started by Alexey and myself. I am largely responsible for the port of python to mingw-w64.

Where on earth did i say I'd remove pip? Please stop building up strawmen here.

**Comment by mingwandroid** [comment]
Fact is pip has no abi atring to distinguish msvc from gcc and that makes it terribly terribly broken for us. Please, if you or anyone who keeps mentioning pip as good-for-msys2 really care so much about pip them please submit patches upstream so we can have an ABI string.

At that point some of my technical objections to pip vanish..

**Comment by naveen521kk** [comment]
It looks like [python2-pip](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-python2-pip/PKGBUILD) uses vendored dependencies. I think soon python3-pip will also do so if it doesn't create any problems.

**Comment by naveen521kk** [comment]
This can be close. Fixed in #7919
